spharpy/spherical.py

```diff
@@ -1,104 +1,585 @@
+
+    @property
+    def basis_type(self):
+        return self._basis_type
+
+    @basis_type.setter
+    def basis_type(self, value):
+        assert value in ["complex", "real"], (
+            "Invalid basis type, currently only 'complex' and"
+            " 'real' "
+            "are supported"
+        )
+        self._basis_type = value
+
+    @property
+    def inverse_transform(self):
+        return self._inverse_transform
+
+    @inverse_transform.setter
+    def inverse_transform(self, value):
+        assert value in ["pseudo_inverse", "quadrature", None], (
+            "Invalid inverse transform type, "
+            "currently only 'pseudo_inverse' "
+            "and 'quadrature' are supported"
+        )
+        self._inverse_transform = value
+
+    @property
+    def channel_convention(self):
+        return self._channel_convention
+
+    @channel_convention.setter
+    def channel_convention(self, value):
+        assert value in ["acn", "fuma"], (
+            "Invalid channel convention, currently only 'acn' "
+            "and 'fuma' are supported"
+        )
+        self._channel_convention = value
+
+    @property
+    def normalization(self):
+        return self._normalization
+
+    @normalization.setter
+    def normalization(self, value):
+        assert value in ["n3d", "maxN", "sn3d"], (
+            "Invalid normalization, "
+            "currently only 'n3d', 'maxN', 'sn3d' are "
+            "supported"
+        )
+        self._normalization = value
+
+    @property
+    def basis(self):
+        if self._basis is None:
+            self._compute_basis()
+        self._check_properties(compute='basis')
+        return self._basis
+
+    @property
+    def basis_gradient_theta(self):
+        if self._basis_gradient_theta is None:
+            self._compute_basis_gradient()
+        self._check_properties(compute='basis_gradient')
+        return self._basis_gradient_theta
+
+    @property
+    def basis_gradient_phi(self):
+        if self._basis_gradient_phi is None:
+            self._compute_basis_gradient()
+        self._check_properties(compute='basis_gradient')
+        return self._basis_gradient_phi
+
+    # @lru_cache(maxsize=128)
+    def _compute_basis(self):
+        """
+        Compute the basis matrix for the SphericalHarmonics class.
+        """
+        logger.info("Computing basis matrix for n_max=%d", self.n_max)
+        if self.basis_type == "complex":
+            self.basis = spherical_harmonic_basis(
+                self.n_max, self.coords,
+                self.normalization, self.channel_convention
+            )
+        elif self.basis_type == "real":
+            self.basis = spherical_harmonic_basis_real(
+                self.n_max, self.coords,
+                self.normalization, self.channel_convention
+            )
+        else:
+            raise ValueError(
+                "Invalid signal type, should be either 'complex' or 'real'"
+            )
```

Comment:
Could be simplified to sth like this.

```suggestion
        if self.basis_type == "complex":
            function = spherical_harmonic_basis
        elif self.basis_type == "real":
            function = spherical_harmonic_basis_real
        else:
            raise ValueError(
                "Invalid signal type, should be either 'complex' or 'real'"
            )
        self.basis = function(
                self.n_max, self.coords,
                self.normalization, self.channel_convention)
```